From 1a43f32d4fa878d3accb23345de7491df81db86e Mon Sep 17 00:00:00 2001
From: dineshsv <dineshsv@ami.com>
Date: Fri, 19 Dec 2025 02:38:53 -0500
Subject: [PATCH] Fixed Postcode Boot count issue.

Tested:

  - Verfied on the AMD congo platform after performing the force warm reboot or
      other power operations boot count is getting increased.

Signed-off-by: dineshsv <dineshsv@ami.com>
---
 inc/post_code.hpp | 45 +++++++++++++++++++++++++--------------------
 1 file changed, 25 insertions(+), 20 deletions(-)

diff --git a/inc/post_code.hpp b/inc/post_code.hpp
index 5ac68a7..b6749cb 100644
--- a/inc/post_code.hpp
+++ b/inc/post_code.hpp
@@ -100,28 +100,31 @@ struct PostCode : sdbusplus::server::object_t<post_code, delete_all>
                 std::string intfName;
                 std::map<std::string, std::variant<std::string>> msgData;
                 msg.read(intfName, msgData);
-                // Check if it was the Value property that changed.
-                auto valPropMap = msgData.find("CurrentHostState");
-                if (valPropMap != msgData.end())
+		auto currentHostStateProp = msgData.find("CurrentHostState");
+                if (currentHostStateProp == msgData.end())
+                {
+                    return;
+                }
+
+                if (!std::holds_alternative<std::string>(
+                                        currentHostStateProp->second))
                 {
-                    StateServer::Host::HostState currentHostState =
-                        StateServer::Host::convertHostStateFromString(
-                            std::get<std::string>(valPropMap->second));
-                    if (currentHostState == StateServer::Host::HostState::Off)
-                    {
-                        if (this->postCodes.empty())
-                        {
-                            std::cerr
-                                << "HostState changed to OFF. Empty "
-                                   "postcode log, keep boot cycle at "
-                                << this->currentBootCycleIndex << std::endl;
-                        }
-                        else
-                        {
-                            this->postCodes.clear();
-                        }
-                    }
+                    return;
                 }
+		auto newHostState =
+                    StateServer::Host::convertHostStateFromString(
+                        std::get<std::string>(currentHostStateProp->second));
+                
+		if (newHostState ==
+                        StateServer::Host::HostState::Running &&
+                    previousHostState !=
+                        StateServer::Host::HostState::Running)
+                {
+                    this->postCodes.clear();
+                }
+
+                previousHostState = newHostState;
+
             })
     {
         phosphor::logging::log<phosphor::logging::level::INFO>(
@@ -154,6 +157,8 @@ struct PostCode : sdbusplus::server::object_t<post_code, delete_all>
     std::map<uint64_t, postcode_t> postCodes;
     fs::path postCodeListPath;
     uint16_t currentBootCycleIndex = 0;
+    StateServer::Host::HostState previousHostState =
+        StateServer::Host::HostState::Off;
     sdbusplus::bus::match_t propertiesChangedSignalRaw;
     sdbusplus::bus::match_t propertiesChangedSignalCurrentHostState;
 
-- 
2.34.1

